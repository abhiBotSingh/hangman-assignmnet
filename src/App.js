import React, { useState, useEffect } from 'react';
import HangmanContext from './context/HangmanContext';
import axios from "axios";

import Title from './components/Title';
import Man from './components/Man';
import WrongAlphabets from './components/WrongAlphabets';
import Word from './components/Word';
import Popup from './components/Popup';
import Notification from './components/Notification';

import { getRandomIndex, showNotification as show } from './helpers/helpers';

import './App.css';

const words = ['temporary', 'wizard', 'programmer', 'teamwork', 'python'];
const hints = [
  'something that is not permanent',
  'someone who is able to perform magic',
  'someone who writes programs',
  'working together',
  'a kind of snake'
];
let initialIndex = getRandomIndex(words.length);
let selectedWord = words[initialIndex];
let hint = hints[initialIndex];

function App() {
  const [playable, setPlayable] = useState(true);
  const [correctAlphabets, setCorrectAlphabets] = useState([]);
  const [wrongAlphabets, setWrongAlphabets] = useState([]);
  const [showNotification, setShowNotification] = useState(false);
  const [word, setWord] = useState("");

  useEffect(() => {
    async function getRandomWord() {
      await axios.get("https://random-word-api.herokuapp.com/word?number=1&swear=0").then((res) => {
        const randomWord = res.data[0];
        console.log(randomWord);
        setWord(randomWord);
      });
    }
    getRandomWord();
  }, []);

  useEffect(() => {
    const handleKeyPress = (event) => {
      const { key, keyCode } = event;
      if (playable && keyCode >= 65 && keyCode <= 90) {
        const letter = key.toLowerCase();
        if (selectedWord.includes(letter)) {
          if (!correctAlphabets.includes(letter)) {
            setCorrectAlphabets(currentAlphabet => [...currentAlphabet, letter]);
          } else {
            show(setShowNotification);
          }
        } else {
          if (!wrongAlphabets.includes(letter)) {
            setWrongAlphabets(currentAlphabet => [...currentAlphabet, letter]);
          } else {
            show(setShowNotification);
          }
        }
      }
    }
    window.addEventListener('keydown', handleKeyPress);

    return () => window.removeEventListener('keydown', handleKeyPress);
  }, [correctAlphabets, wrongAlphabets, playable]);

  function playAgain() {
    setPlayable(true);
    setCorrectAlphabets([]);
    setWrongAlphabets([]);

    const random = getRandomIndex(words.length);
    selectedWord = words[random];
    hint = hints[random];
  }

  return (
    <>
      <Title />
      <div className="game-container">
        <HangmanContext.Provider value={wrongAlphabets}>
          <Man />
          <WrongAlphabets />
        </HangmanContext.Provider>

        <HangmanContext.Provider value={{ selectedWord, correctAlphabets, hint }}>
          <Word />
        </HangmanContext.Provider>
        <p className="hint">HINT - {hint}</p>
      </div>
      <HangmanContext.Provider value={{ correctAlphabets, wrongAlphabets, selectedWord, setPlayable, playAgain }}>
        <Popup />
      </HangmanContext.Provider>

      <HangmanContext.Provider value={showNotification}>
        <Notification />
      </HangmanContext.Provider >
    </>
  );
}

export default App;
