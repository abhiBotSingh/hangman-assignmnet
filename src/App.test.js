import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import { act } from "react-dom/test-utils";
import App from "./App";
import axios from "axios";

jest.mock("axios");

const mockWord = "mock"
const testId = "hangman";

describe("Hangman game tests", () => {
    test("Hangman game is rendered properly", async () => {
        axios.get = jest
            .fn()
            .mockResolvedValueOnce({ data: [mockWord] })
            .mockResolvedValueOnce({
                data: [
                    {
                        "meanings": [{
                            "definitions": [{
                                "definition": "mock hint"
                            }]
                        }]
                    }
                ]
            });

        await act(async () => {
            const { asFragment } = render(<App />);
            expect(asFragment()).toMatchSnapshot();
        });
        expect(axios.get).toHaveBeenCalledTimes(2);
    });

    test("Hangman body parts appear one-by-one when an incorrect alphabet is pressed", async () => {
        axios.get = jest
            .fn()
            .mockResolvedValueOnce({ data: [mockWord] })
            .mockResolvedValueOnce({
                data: [
                    {
                        "meanings": [{
                            "definitions": [{
                                "definition": "mock hint"
                            }]
                        }]
                    }
                ]
            });

        await act(async () => {
            render(<App />);
        });

        const heading = screen.getByTestId(`${testId}-heading`);
        expect(heading).toBeVisible;

        //Checking whether the head of man appears on 1st incorrect key press
        fireEvent.keyDown(heading,
            { key: 'a', keyCode: 65, code: "keyA" }
        );
        const hangmanHead = screen.getByTestId(`${testId}-head`);
        expect(hangmanHead).toBeVisible;

        //Checking whether the body of man appears on 2nd incorrect key press
        fireEvent.keyDown(heading,
            { key: 'b', keyCode: 66, code: "keyB" }
        );
        const hangmanBody = screen.getByTestId(`${testId}-body`);
        expect(hangmanBody).toBeVisible;
    });

    test("Success popup appears when the player wins the game", async () => {
        axios.get = jest
            .fn()
            .mockResolvedValueOnce({ data: ["z"] })
            .mockResolvedValueOnce({
                data: [
                    {
                        "meanings": [{
                            "definitions": [{
                                "definition": "mock hint"
                            }]
                        }]
                    }
                ]
            });

        await act(async () => {
            render(<App />);
        });
        const heading = screen.getByTestId(`${testId}-heading`);

        fireEvent.keyDown(heading,
            { key: 'z', keyCode: 90, code: "keyZ" }
        );

        const successMessage = screen.queryByText("You won! Congratulations!");
        expect(successMessage).toHaveLength;
    });

    test("Play again button appears when the player wins the game", async () => {
        axios.get = jest
            .fn()
            .mockResolvedValueOnce({ data: [mockWord] })
            .mockResolvedValueOnce({
                data: [
                    {
                        "meanings": [{
                            "definitions": [{
                                "definition": "mock hint"
                            }]
                        }]
                    }
                ]
            });

        await act(async () => {
            render(<App />);
        });
        const heading = screen.getByTestId(`${testId}-heading`);

        fireEvent.keyDown(heading,
            { key: 'z', keyCode: 90, code: "keyZ" }
        );

        const playAgainButton = screen.getByTestId(`${testId}-play-again`);
        expect(playAgainButton).toBeVisible;
    });

    test("A notification appears when an alphabet that has been pressed already, is pressed again", async () => {
        axios.get = jest
            .fn()
            .mockResolvedValueOnce({ data: [mockWord] })
            .mockResolvedValueOnce({
                data: [
                    {
                        "meanings": [{
                            "definitions": [{
                                "definition": "mock hint"
                            }]
                        }]
                    }
                ]
            });

        await act(async () => {
            render(<App />);
        });
        const heading = screen.getByTestId(`${testId}-heading`);

        fireEvent.keyDown(heading,
            { key: 'z', keyCode: 90, code: "keyZ" }
        );

        const letterPressedAgain = screen.queryByText("You have already entered this letter");
        expect(letterPressedAgain).toHaveLength;
    });

    test("Player loses the game if they enter 6 incorrect alphabets", async () => {
        axios.get = jest
            .fn()
            .mockResolvedValueOnce({ data: [mockWord] })
            .mockResolvedValueOnce({
                data: [
                    {
                        "meanings": [{
                            "definitions": [{
                                "definition": "mock hint"
                            }]
                        }]
                    }
                ]
            });

        await act(async () => {
            render(<App />);
        });
        const heading = screen.getByTestId(`${testId}-heading`);

        //Entering 6 incorrect alphabets
        fireEvent.keyDown(heading,
            { key: 'a', keyCode: 65, code: "keyA" }
        );
        fireEvent.keyDown(heading,
            { key: 'b', keyCode: 66, code: "keyB" }
        );
        fireEvent.keyDown(heading,
            { key: 'c', keyCode: 67, code: "keyC" }
        );
        fireEvent.keyDown(heading,
            { key: 'd', keyCode: 68, code: "keyD" }
        );
        fireEvent.keyDown(heading,
            { key: 'e', keyCode: 69, code: "keyE" }
        );
        fireEvent.keyDown(heading,
            { key: 'f', keyCode: 70, code: "keyF" }
        );

        const gameLoseMessage = screen.queryByText(`...the word was: ${mockWord}`);
        expect(gameLoseMessage).toBeVisible;
    });
});