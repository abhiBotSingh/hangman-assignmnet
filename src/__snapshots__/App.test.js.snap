// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hangman game tests Hangman game is rendered properly 1`] = `
<DocumentFragment>
  <h1
    data-testid="hangman-heading"
  >
    Hangman
  </h1>
  <p>
    Find the hidden word - Enter a letter
  </p>
  <p
    class="hint"
  >
    HINT - 
  </p>
  <div
    class="game-container"
  >
    <svg
      class="man-container"
      height="250"
      width="200"
    >
      <line
        x1="60"
        x2="140"
        y1="20"
        y2="20"
      />
      <line
        x1="140"
        x2="140"
        y1="20"
        y2="50"
      />
      <line
        x1="60"
        x2="60"
        y1="20"
        y2="230"
      />
      <line
        x1="20"
        x2="100"
        y1="230"
        y2="230"
      />
    </svg>
    <div
      class="wrong-alphabets-container"
    >
      <div />
    </div>
    <div
      class="word"
    />
  </div>
  <div
    class="notification-container "
  >
    <p>
      You have already entered this letter
    </p>
  </div>
</DocumentFragment>
`;
